kube/src/api/streaming.rs

```diff
@@ -0,0 +1,274 @@
+// Code to support streaming resources (attach/exec/portforward)
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    sync::Mutex,
+    task::{Context, Poll, Waker},
+};
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::Status;
+
+use async_tungstenite::{
+    tokio::ConnectStream,
+    tungstenite::{self as ws, Message},
+    WebSocketStream,
+};
+use futures::{future::Either, SinkExt, Stream, StreamExt, TryStreamExt};
+use futures_util::future::select;
+use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt, DuplexStream};
+use tokio_util::codec;
+
+// Internal state of an attached process
+struct AttachedProcessState {
+    waker: Option<Waker>,
+    finished: bool,
+    status: Option<Status>,
+    stdin_writer: Option<DuplexStream>,
+    stdout_reader: Option<DuplexStream>,
+    stderr_reader: Option<DuplexStream>,
+}
+
+/// Represents an attached process in a container for `attach` and `exec`.
+///
+/// Provides access to stdin/stdout/stderr if attached.
+/// Resolves when the connection terminates with an optional [`Status`].
+///
+/// [`Status`]: k8s_openapi::apimachinery::pkg::apis::meta::v1::Status
+pub struct AttachedProcess {
+    has_stdin: bool,
+    has_stdout: bool,
+    has_stderr: bool,
+    state: Arc<Mutex<AttachedProcessState>>,
+}
+
+impl AttachedProcess {
+    pub(crate) fn new(
+        stream: WebSocketStream<ConnectStream>,
+        stdin: bool,
+        stdout: bool,
+        stderr: bool,
+    ) -> Self {
+        // To simplify the implementation, always create a pipe for stdin.
+        // The caller does not have access to it unless they had requested.
+        // REVIEW Make internal buffer size configurable?
+        let (stdin_writer, stdin_reader) = tokio::io::duplex(1024);
+        let (stdout_writer, stdout_reader) = if stdout {
+            let (w, r) = tokio::io::duplex(1024);
+            (Some(w), Some(r))
+        } else {
+            (None, None)
+        };
+        let (stderr_writer, stderr_reader) = if stderr {
+            let (w, r) = tokio::io::duplex(1024);
+            (Some(w), Some(r))
+        } else {
+            (None, None)
+        };
+
+        let state = Arc::new(Mutex::new(AttachedProcessState {
+            waker: None,
+            finished: false,
+            status: None,
+            stdin_writer: Some(stdin_writer),
+            stdout_reader,
+            stderr_reader,
+        }));
+        let shared_state = state.clone();
+        tokio::spawn(async move {
+            let status = start(stream, stdin_reader, stdout_writer, stderr_writer)
+                .await
+                .unwrap();
+
+            let mut shared = shared_state.lock().unwrap();
+            shared.finished = true;
+            shared.status = status;
+            if let Some(waker) = shared.waker.take() {
+                waker.wake()
```

Comment:
it looks like we are creating an empty `AttachedProcessState`, filling it in in a `spawn`, then fully awaiting it there, and notifying when it's ready, then implementing Future for the other temporary object we return.

If feels like if we forced the user to pass in this stream, we would not have to do this, and we could just await the `start` call?

Reply:
When `AttachedProcess` is created, `AttachedProcessState` contains the following:

- `stdin_writer`: Writer half of the `stdin` pipe. Always present to simplify implementation, but only accessible if `stdin: true`
- `stdout_reader`: Reader half of the `stdout` pipe if `stdout: true`
- `stdout_reader`: Reader half of the `stderr` pipe if `stderr: true`

These are used to create writer and streams for users.

> If feels like if we forced the user to pass in this stream

Users don't pass in anything to `AttachedProcess`.

To write to `stdin`, users take the writer out and write to it.

```rust
let mut stdin_writer = attached.stdin().unwrap();
stdin_writer.write(b"foo\n");
```

To read from `stdout`/`stderr`, users take streams out:

```rust
let mut stdout_stream = attached.stdout().unwrap();
let next_stdout = stdout_stream.next().await?;
// stdout_stream.forward(sink);
// select(stdout_stream, stderr_stream); // combine outputs
```

`start` is misleading, but it's to "start handling messages". The connection is already established and execution has already started in the pod. I did it this way to start handling messages as soon as possible while allowing users to access to writer/streams.

Reply:
Thanks, this clarifies it a lot for me. I had misunderstood it earlier today from a quicker reading.

I think the crucial point for me was that the `AttachedProcess` struct (which has a lot of similarly named structs) has the most crucial methods; the ones that let you interact with the streams. The `stdout`, `stdin` and `stderr` methods there should probably have some extra lines of documentation to describe what people can do with them and what they correspond to (in terms of input booleans and channels and the ws side).